File: Python/programmers/lv2/전력망을 둘로 나누기.py
def solution(n, wires):
    global connect

    connect = [[False] * (n + 1) for _ in range(n + 1)]
    for (v1, v2) in wires:
        connect[v1][v2] = True
        connect[v2][v1] = True

    answer = 987654321

    for (v1, v2) in wires:
        global size
        size = 0

        connect[v1][v2] = False
        connect[v2][v1] = False

        dfs(n, 1, [False] * (n + 1))
        connect[v1][v2] = True
        connect[v2][v1] = True

        answer = min(answer, abs(size - (n - size)))

    return answer


def dfs(n, cur, visited):
    global connect
    global size
    size += 1
    visited[cur] = True

    for next_node in range(1, n + 1):
        if connect[cur][next_node] == True and visited[next_node] == False:
            dfs(n, next_node, visited)

File: kotlin_java/src/programmers/lv2/전력망을_둘로_나누기/전력망을 둘로 나누기.kt
package programmers.lv2.전력망을_둘로_나누기

import java.util.*
import kotlin.math.*

lateinit var connected: Array<Array<Boolean>>
lateinit var visited: Array<Boolean>
var depth = 0
var blockSize = 0

fun solution(n: Int, wires: Array<IntArray>): Int {
    var best = Int.MAX_VALUE

    connected = Array(n+1){Array(n+1){false}}

    for(wire in wires) {
        connected[wire[0]][wire[1]] = true
        connected[wire[1]][wire[0]] = true
    }

    for(wire in wires) {
        blockSize = 0
        depth = 1
        visited = Array(n + 1) { false }

        connected[wire[0]][wire[1]] = false
        connected[wire[1]][wire[0]] = false


        visited[1] = true

        dfs(n, 1)
        best = min(best, (abs((blockSize) - abs(n - blockSize))))

        connected[wire[0]][wire[1]] = true
        connected[wire[1]][wire[0]] = true
    }

    return best
}


fun dfs(n:Int, idx:Int) {

    blockSize = max(blockSize, depth)

    for(i in 1 .. n) {
        if(connected[idx][i] && !visited[i]) {
            visited[i] = true
            depth++
            dfs(n,i)
        }
    }
}